### PR TITLE
Update CLI to respect fix applicability

### PR DESCRIPTION
**Pull request by zanieb**
Rebase of https://github.com/astral-sh/ruff/pull/5119 authored by @evanrittenhouse with additional refinements.

## Changes

- Adds `--unsafe-fixes` / `--no-unsafe-fixes` flags to `ruff check`
    - Violations with unsafe fixes are not shown as fixable unless opted-in
- Fix applicability is respected now
    - `Applicability::Never` fixes are no longer applied
    - `Applicability::Sometimes` fixes require opt-in
    - `Applicability::Always` fixes are unchanged
 - Hints for availability of `--unsafe-fixes` added to `ruff check` output

## Examples

Check hints at hidden unsafe fixes
```
❯ ruff check example.py --no-cache --select F601,W292
example.py:1:14: F601 Dictionary key literal `'a'` repeated
example.py:2:15: W292 [*] No newline at end of file
Found 2 errors.
[*] 1 fixable with the `--fix` option (1 hidden fix can be enabled with the `--unsafe-fixes` option).
```

We could add an indicator for which violations have hidden fixes in the future.

Check treats unsafe fixes as applicable with opt-in
```
❯ ruff check example.py --no-cache --select F601,W292 --unsafe-fixes
example.py:1:14: F601 [*] Dictionary key literal `'a'` repeated
example.py:2:15: W292 [*] No newline at end of file
Found 2 errors.
[*] 2 fixable with the --fix option.
```

Also can be enabled in the config file

```
❯ cat ruff.toml
unsafe-fixes = true
```

And opted-out per invocation

```
❯ ruff check example.py --no-cache --select F601,W292 --no-unsafe-fixes
example.py:1:14: F601 Dictionary key literal `'a'` repeated
example.py:2:15: W292 [*] No newline at end of file
Found 2 errors.
[*] 1 fixable with the `--fix` option (1 hidden fix can be enabled with the `--unsafe-fixes` option).
```

Diff does not include unsafe fixes
```
❯ ruff check example.py --no-cache --select F601,W292 --diff
--- example.py
+++ example.py
@@ -1,2 +1,2 @@
 x = {'a': 1, 'a': 1}
-print(('foo'))
+print(('foo'))
\ No newline at end of file

Would fix 1 error.
```

Unless there is opt-in
```
❯ ruff check example.py --no-cache --select F601,W292 --diff --unsafe-fixes
--- example.py
+++ example.py
@@ -1,2 +1,2 @@
-x = {'a': 1}
-print(('foo'))
+x = {'a': 1, 'a': 1}
+print(('foo'))
\ No newline at end of file

Would fix 2 errors.
```

https://github.com/astral-sh/ruff/pull/7790 will improve the diff messages following this pull request

Similarly, `--fix` and `--fix-only` require the `--unsafe-fixes` flag to apply unsafe fixes.

## Related

Replaces #5119
Closes https://github.com/astral-sh/ruff/issues/4185
Closes https://github.com/astral-sh/ruff/issues/7214
Closes https://github.com/astral-sh/ruff/issues/4845
Closes https://github.com/astral-sh/ruff/issues/3863
Addresses https://github.com/astral-sh/ruff/issues/6835
Addresses https://github.com/astral-sh/ruff/issues/7019
Needs follow-up https://github.com/astral-sh/ruff/issues/6962
Needs follow-up https://github.com/astral-sh/ruff/issues/4845
Needs follow-up https://github.com/astral-sh/ruff/issues/7436
Needs follow-up https://github.com/astral-sh/ruff/issues/7025
Needs follow-up https://github.com/astral-sh/ruff/issues/6434
Follow-up #7790 
Follow-up https://github.com/astral-sh/ruff/pull/7792

### CODE REVIEW

File: crates/ruff_cli/src/args.rs [hunk 1]
```diff
@@ -72,15 +72,17 @@ pub enum Command {
...
-#[allow(clippy::struct_excessive_bools)]
 pub struct CheckCommand {
     /// List of files or directories to check.
     pub files: Vec<PathBuf>,
     /// Attempt to automatically fix lint violations.
     /// Use `--no-fix` to disable.
     #[arg(long, overrides_with("no_fix"))]
     fix: bool,
-    #[clap(long, overrides_with("fix"), hide = true)]
+    /// Attempt to automatically fix both automatic and suggested lint violations.
```
Comment (zanieb) [inline crates/ruff_cli/src/args.rs:82]:
```suggestion
    /// Attempt to fix both automatic and suggested lint violations.
```

File: crates/ruff_cli/src/diagnostics.rs [hunk 2]
```diff
@@ -241,96 +241,110 @@ pub(crate) fn lint_path(
...
+                                SourceKind::IpyNotebook(dest_notebook) => {
+                                    // We need to load the notebook again, since we might've
+                                    // mutated it.
+                                    let src_notebook = source_kind.as_ipy_notebook().unwrap();
+                                    let mut stdout = io::stdout().lock();
+                                    for ((idx, src_cell), dest_cell) in src_notebook
+                                        .cells()
+                                        .iter()
+                                        .enumerate()
+                                        .zip(dest_notebook.cells().iter())
```
Comment (zanieb) [inline crates/ruff_cli/src/diagnostics.rs:293]: Revert this change

File: crates/ruff_cli/src/diagnostics.rs [hunk 3]
```diff
@@ -392,7 +406,7 @@ pub(crate) fn lint_stdin(
     };
 
     // Extract the sources from the file.
-    let source_kind = match SourceKind::from_source_code(contents, source_type) {
+    let source_kind = match SourceKind::from_source_code(contents.clone(), source_type) {
```
Comment (zanieb) [inline crates/ruff_cli/src/diagnostics.rs:409]: Investigate avoiding this clone

File: crates/ruff_cli/src/diagnostics.rs [hunk 4]
```diff
@@ -407,49 +421,70 @@ pub(crate) fn lint_stdin(
...
+                noqa,
+                &settings.linter,
+                &source_kind,
+                source_type,
+                fix_suggested,
+            ) {
+                match fix_mode {
+                    flags::FixMode::Apply(_) => {
+                        // Write the contents to stdout, regardless of whether any errors were fixed.
+                        io::stdout().write_all(transformed.source_code().as_bytes())?;
```
Comment (zanieb) [inline crates/ruff_cli/src/diagnostics.rs:442]: Why the change in `transformed.write`?

File: crates/ruff_cli/src/lib.rs [hunk 5]
```diff
@@ -237,16 +238,28 @@ pub fn check(args: CheckCommand, log_level: LogLevel) -> Result<ExitStatus> {
...
     // - If `--fix` or `--fix-only` is set, always apply fixes to the filesystem (or
-    //   print them to stdout, if we're reading from stdin).
+    //   print them to stdout, if we're reading from stdin) for [`Applicability::Automatic`] rules
+    //   only.
     // - If `--diff` or `--fix-only` are set, don't print any violations (only
-    //   fixes).
+    //   fixes) for [`Applicability::Automatic`] rules only.
+    // - If `--fix--suggested` is set, the above rules will apply to both [`Applicability::Suggested`] and
+    //   [`Applicability::Automatic`] fixes.
+    // TODO: can't fix this until @zanieb changes the CLI
```
Comment (zanieb) [inline crates/ruff_cli/src/lib.rs:247]: What does this refer to?

File: crates/ruff_cli/src/printer.rs [hunk 6]
```diff
@@ -191,9 +184,9 @@ impl Printer {
             }
             SerializationFormat::Text => {
                 TextEmitter::default()
-                    .with_show_fix_status(show_fix_status(self.fix_mode))
-                    .with_show_fix_diff(self.flags.intersects(Flags::SHOW_FIX_DIFF))
-                    .with_show_source(self.flags.intersects(Flags::SHOW_SOURCE))
+                    .with_show_fix_status(show_fix_status(self.fix_mode, fixables))
+                    .with_show_fix_diff(self.flags.contains(Flags::SHOW_FIX_DIFF))
```
Comment (zanieb) [inline crates/ruff_cli/src/printer.rs:188]: Why `contains` instead of `intersects`?

File: crates/ruff_cli/tests/integration_test.rs [hunk 7]
```diff
@@ -869,3 +797,232 @@ fn check_input_from_argfile() -> Result<()> {
 
     Ok(())
 }
+
+#[test]
+fn displays_fix_applicability_levels() -> Result<()> {
+    // `--fix` should only apply automatic fixes, but should tell the user about `--fix --unautomatic` if
+    // there are remaining unautomatic fixes.
+    // TODO: this should be a failure but we don't have a way to track that
```
Comment (zanieb) [inline crates/ruff_cli/tests/integration_test.rs:805]: This seems incorrect, `success: false` seems to track this.

File: crates/ruff_diagnostics/src/fix.rs [hunk 8]
```diff
@@ -5,27 +5,30 @@ use ruff_text_size::{Ranged, TextSize};
...
-
     /// The fix has a good chance of being incorrect or the code be incomplete.
     /// The fix may result in invalid code if it is applied.
     /// The fix should only be manually applied by the user.
     Manual,
 
-    /// The applicability of the fix is unknown.
+    /// The applicability of the fix is unknown or not relevant.
     #[default]
     Unspecified,
```
Comment (zanieb) [inline crates/ruff_diagnostics/src/fix.rs:22]: We should probably drop `Unspecified` as it's only there for transition purposes?

Comment (zanieb) [reply crates/ruff_diagnostics/src/fix.rs:22]: Let's do this in a follow-up pull request to keep the diff small.

**Comment by zanieb** [comment]
Rebased @evanrittenhouse to be the author of the first commit — I am leaning towards using this pull request since I've got details in the body.

File: crates/ruff_cli/src/diagnostics.rs (hunk 3 above)
Comment (zanieb) [reply crates/ruff_cli/src/diagnostics.rs]: I avoided this but now pass a reference?

File: crates/ruff_cli/src/args.rs [hunk 9]
```diff
@@ -76,11 +76,14 @@ pub enum Command {
 pub struct CheckCommand {
     /// List of files or directories to check.
     pub files: Vec<PathBuf>,
-    /// Attempt to automatically fix lint violations.
-    /// Use `--no-fix` to disable.
+    /// Apply automatic fixes to resolve lint violations.
```
Comment (konstin) [inline crates/ruff_cli/src/args.rs:79]: Personal preference:

```suggestion
    /// Apply fixes to resolve lint violations.
```

File: crates/ruff_cli/src/diagnostics.rs [hunk 10]
```diff
@@ -241,96 +241,109 @@ pub(crate) fn lint_path(
...
+                fixed,
+            }) = lint_fix(
+                path,
+                package,
+                noqa,
+                settings,
+                &source_kind,
+                source_type,
+                fix_suggested,
+            ) {
```
Comment (konstin) [inline crates/ruff_cli/src/diagnostics.rs:258]: nit: could you give this a temporary variable

File: crates/ruff_cli/src/diagnostics.rs (hunk 4 above)
Comment (konstin) [reply crates/ruff_cli/src/diagnostics.rs:441]: Either looks fine to me

File: crates/ruff_cli/src/printer.rs [hunk 11]
```diff
@@ -439,3 +434,71 @@ fn print_fix_summary(writer: &mut dyn Write, fixed: &FxHashMap<String, FixTable>
...
+}
+
+impl<'a> FixableStatistics<'a> {
+    fn new(diagnostics: &Diagnostics, apply_suggested: &'a SuggestedFixes) -> Self {
+        let mut automatic = 0;
+        let mut suggested = 0;
+
+        for message in &diagnostics.messages {
+            if let Some(fix) = &message.fix {
+                if fix.applicability() == Applicability::Suggested {
```
Comment (konstin) [inline crates/ruff_cli/src/printer.rs:452]: i'd match here

File: crates/ruff_cli/src/args.rs [hunk 12]
```diff
@@ -76,11 +76,14 @@ pub enum Command {
...
+    /// Use `--no-fix` to disable or `--fix-suggested` to include suggested fixes.
     #[arg(long, overrides_with("no_fix"))]
     fix: bool,
-    #[clap(long, overrides_with("fix"), hide = true)]
+    /// Apply automatic and suggested fixes to resolve lint violations.
+    #[arg(long, overrides_with_all(["fix", "no_fix"]))]
+    fix_suggested: bool,
```
Comment (charliermarsh) [inline crates/ruff_cli/src/args.rs:85]: What was the reasoning behind using `--fix-suggested` instead of `--unsafe`? Could we also consider `--suggested` or `--with-suggested` or `--include-suggested`? I don't want to derail the PR, but the confusion for me (IIRC) is that it's similar to `--fix` but conceptually independent.

For example, you're now supposed to do `--fix --fix-suggested` (`--fix-suggested` alone as no effect, IIUC), but `--fix-suggested` feels like it should imply `--fix`.

Similarly, you need to do `--diff --fix-suggested` whereas `--diff --fix` would be invalid.

File: crates/ruff_cli/tests/integration_test.rs [hunk 13]
```diff
@@ -861,11 +789,223 @@ fn check_input_from_argfile() -> Result<()> {
...
+            .pass_stdin("x = {'a': 1, 'a': 1}\nprint(('foo'))\n"),
+            @r###"
+    success: false
+    exit_code: 1
+    ----- stdout -----
+    -:1:14: F601 [**] Dictionary key literal `'a'` repeated
+    -:2:7: UP034 [*] Avoid extraneous parentheses
+    Found 2 errors.
+    [*] 1 fixable with the --fix option.
+    [**] 1 potentially fixable with the --fix-suggested option.
```
Comment (konstin) [inline crates/ruff_cli/tests/integration_test.rs:896]: That's confusing, the user just passed `--fix-suggested` and now we're telling them to use `--fix-suggested`, also shouldn't we get the fixed source back?

File: crates/ruff_diagnostics/src/fix.rs [hunk 14]
```diff
@@ -5,27 +5,40 @@ use ruff_text_size::{Ranged, TextSize};
...
+    /// The fix is definitely what the user intended, or maintains the exact meaning of the code.
+    /// This fix should be automatically applied.
+    Automatic,
+}
+
+impl Applicability {
+    pub fn symbol(&self) -> &'static str {
+        match self {
+            Self::Automatic => "*",
+            Self::Suggested => "**",
```
Comment (charliermarsh) [inline crates/ruff_diagnostics/src/fix.rs:38]: Having played around with it in the terminal, I think I prefer `^` over `**`, it feels cleaner to me. But I recognize that it's somewhat subjective.

File: crates/ruff_cli/src/args.rs (hunk 12 above)
Comment (zanieb) [reply crates/ruff_cli/src/args.rs:85]: Ah gee... I intended for `--fix-suggested` to imply` --fix` but that gets weird with `--diff` which I guess is why Evan went with this approach?

There's lots of discussion elsewhere about why not `--unsafe` but basically the idea was that we wanted consistency between the internal "Suggested" level (which is also going to be displayed in the output eventually) and the user facing concept.

File: crates/ruff_diagnostics/src/fix.rs [hunk 15]
```diff
@@ -5,27 +5,40 @@ use ruff_text_size::{Ranged, TextSize};
...
     #[default]
     Unspecified,
+
+    /// The fix may be what the user intended, but it is uncertain.
+    /// The fix should result in valid code if it is applied.
+    /// The fix can be applied with user opt-in.
+    Suggested,
+
+    /// The fix is definitely what the user intended, or maintains the exact meaning of the code.
+    /// This fix should be automatically applied.
```
Comment (konstin) [inline crates/ruff_diagnostics/src/fix.rs:30]: nit: Grammar and consistency in these docstrings

File: crates/ruff_cli/src/args.rs (hunk 12 above)
Comment (zanieb) [reply crates/ruff_cli/src/args.rs:85]: My qualms with `--suggested` / `--with-suggested` / `--include-suggested` is that it is not clear that they refer to fixes specifically.

File: crates/ruff_linter/src/message/text.rs [hunk 16]
```diff
@@ -141,11 +141,13 @@ impl Display for RuleCodeAndBody<'_> {
         let kind = &self.message.kind;
 
         if self.show_fix_status && self.message.fix.is_some() {
+            let indicator = self.message.fix.as_ref().unwrap().applicability().symbol();
```
Comment (charliermarsh) [inline crates/ruff_linter/src/message/text.rs:144]: Can we use an `if-let` here instead of `.unwrap()`? We already test `.is_some()` above, we should be able to do `if let Some(fix) = message.fix.as_ref()` instead of `.is_some()` + `.unwrap()`.

**Review by charliermarsh — COMMENTED** [review]
I haven't debugged it, but locally, passing `--fix-suggested` is preventing Ruff from applying any fixes for me.

E.g., I have:

```python
getattr(x, "y")

import os
```

And I see:

```
❯ cargo run -p ruff_cli -- check foo.py --select B --fix --fix-suggested -n --isolated
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
     Running `target/debug/ruff check foo.py --select B --fix --fix-suggested -n --isolated`
foo.py:1:1: B009 [^] Do not call `getattr` with a constant attribute value. It is not any safer than normal property access.
Found 1 error.
[^] 1 potentially fixable with the --fix-suggested option.
```

File: crates/ruff_linter/src/message/text.rs (hunk 16 above)
Comment (zanieb) [reply crates/ruff_linter/src/message/text.rs:144]: Yeah I'm actually in the middle of changing this already :)

File: crates/ruff_linter/src/settings/flags.rs [hunk 17]
```diff
@@ -1,8 +1,24 @@
 #[derive(Debug, Copy, Clone, Hash, is_macro::Is)]
 pub enum FixMode {
-    Generate,
+    Generate(SuggestedFixes),
+    Apply(SuggestedFixes),
+    Diff(SuggestedFixes),
+}
```
Comment (konstin) [inline crates/ruff_linter/src/settings/flags.rs:6]: it feels like this should be a struct with two enums, but not critical to merging

File: crates/ruff_cli/tests/integration_test.rs [hunk 18]
```diff
@@ -861,11 +789,223 @@ fn check_input_from_argfile() -> Result<()> {
...
+fn fix_applies_automatic_fixes_only() {
+    // `--fix` should only apply automatic fixes. Since we're runnnig in `stdin` mode, output shouldn't
+    // be printed.
+    assert_cmd_snapshot!(
+        Command::new(get_cargo_bin(BIN_NAME))
+            .args([
+                "-",
+                "--output-format",
+                "text",
+                "--isolated","--no-cache",
```
Comment (konstin) [inline crates/ruff_cli/tests/integration_test.rs:856]: formatting

File: crates/ruff_cli/src/printer.rs [hunk 19]
```diff
@@ -439,3 +434,71 @@ fn print_fix_summary(writer: &mut dyn Write, fixed: &FxHashMap<String, FixTable>
     }
     Ok(())
 }
+
+/// Contains the number of [`Applicability::Automatic`] and [`Applicability::Suggested`] fixes
+struct FixableStatistics<'a> {
```
Comment (charliermarsh) [inline crates/ruff_cli/src/printer.rs:439]: Nit: `#[derive(Debug)]`

**Review by konstin — APPROVED** [review]

File: crates/ruff_cli/src/printer.rs [hunk 20]
```diff
@@ -439,3 +434,71 @@ fn print_fix_summary(writer: &mut dyn Write, fixed: &FxHashMap<String, FixTable>
...
+                "{automatic_prefix} {} fixable with the --fix option.",
+                self.automatic,
+            )
+        } else if self.suggested > 0 {
+            format!(
+                "{suggested_prefix} {} potentially fixable with the --fix-suggested option.",
+                self.suggested
+            )
+        } else {
+            String::new()
```
Comment (charliermarsh) [inline crates/ruff_cli/src/printer.rs:501]: This case combined with `is_empty` feels slightly off to me, since if someone is calling `violation_string` on an empty set, it's probably a mistake? I could see a few alternatives...

1. `fn new` could be `fn try_from` or something like that, and it could return `Option<Self>` (`None` when there are no fixes), so that it's impossible to represent an empty set of fixes.
2. This method could return an `Option` (and `is_empty` could be removed).

(1) feels the most natural to me. What do you think?

File: crates/ruff_diagnostics/src/fix.rs (hunk 14 above)
Comment (charliermarsh) [reply crates/ruff_diagnostics/src/fix.rs:38]: I like it in the terminal too. I think the highlighting we apply helps a bit?

File: crates/ruff_cli/tests/integration_test.rs (hunk 13 above)
Comment (charliermarsh) [reply crates/ruff_cli/tests/integration_test.rs:896]: Hmm yeah, what's going on here? :)

File: crates/ruff_cli/src/lib.rs [hunk 21]
```diff
@@ -236,17 +237,27 @@ pub fn check(args: CheckCommand, log_level: LogLevel) -> Result<ExitStatus> {
...
-    // - If `--fix` or `--fix-only` is set, always apply fixes to the filesystem (or
+    // - If `--fix` or `--fix-only` is set, apply applicable fixes to the filesystem (or
     //   print them to stdout, if we're reading from stdin).
-    // - If `--diff` or `--fix-only` are set, don't print any violations (only
-    //   fixes).
+    // - If `--diff` or `--fix-only` are set, don't print any violations (only applicable fixes)
+    // - By default, applicable fixes only include [`Applicablility::Automatic`], but if
+    //   `--fix--suggested` is set, then [`Applicablility::Suggested`] fixes are included.
+
+    let fix_suggested = if fix_suggested {
```
Comment (charliermarsh) [inline crates/ruff_cli/src/lib.rs:246]: I feel like passing `--fix-suggested` without any of `--fix`, `--diff`, etc. should raise an error on the CLI. Can we enforce that with Clap?

File: crates/ruff_cli/src/args.rs (hunk 12 above)
Comment (evanrittenhouse) [reply crates/ruff_cli/src/args.rs]: Yeah, I went with unsafe originally to expand it to all commands (like --diff --fix-suggested makes less sense than --diff ---unsafe). Maybe --include-unsafe?

Comment (zanieb) [reply crates/ruff_cli/src/args.rs]: I think we're going to go with `--unsafe-fixes` and change the internal names somehow (e.g. #7819)

File: crates/ruff_cli/src/diagnostics.rs [hunk 22]
```diff
@@ -244,8 +246,15 @@ pub(crate) fn lint_path(
             result,
             transformed,
             fixed,
-        }) = lint_fix(path, package, noqa, settings, &source_kind, source_type)
-        {
+        }) = lint_fix(
```
Comment (zanieb) [inline crates/ruff_cli/src/diagnostics.rs:249]: I reverted some changes around here to minimize the diff

File: crates/ruff_cli/src/diagnostics.rs (hunk 10 above)
Comment (zanieb) [reply crates/ruff_cli/src/diagnostics.rs]: I'm not sure what you mean here, but I'd like to address this separately to keep the diff clear (I reverted some of Evan's changes here to keep the diff simpler already).

File: crates/ruff_cli/src/args.rs [hunk 23]
```diff
@@ -76,12 +77,18 @@ pub enum Command {
 pub struct CheckCommand {
     /// List of files or directories to check.
     pub files: Vec<PathBuf>,
-    /// Attempt to automatically fix lint violations.
-    /// Use `--no-fix` to disable.
+    /// Apply fixes to resolve lint violations.
+    /// Use `--no-fix` to disable or `--unsafe-fixes` to include unsafe fixes.
```
Comment (charliermarsh) [inline crates/ruff_cli/src/args.rs:81]: Should we say "potentially unsafe fixes"? (Defer to you.)

File: crates/ruff_cli/src/diagnostics.rs [hunk 24]
```diff
@@ -244,8 +246,15 @@ pub(crate) fn lint_path(
...
+            path,
+            package,
+            noqa,
+            unsafe_fixes,
+            settings,
+            &source_kind,
+            source_type,
+        ) {
```
Comment (charliermarsh) [inline crates/ruff_cli/src/diagnostics.rs:257]: I think Konsti's suggestion (which it's totally up to you whether you take) was to do, like:

```
let var = lint_fix(...);
if let Ok(FixerResult { ... }) = var {
  ....
}
```

...to avoid having a long expanded call on the RHS of the `if-let`. Again, completely up to you.

File: crates/ruff_cli/src/commands/check.rs [hunk 25]
```diff
@@ -119,7 +120,16 @@ pub(crate) fn check(
...
 
-                    lint_path(path, package, &settings.linter, cache, noqa, fix_mode).map_err(|e| {
+                    lint_path(
+                        path,
+                        package,
+                        &settings.linter,
+                        cache,
+                        noqa,
+                        fix_mode,
+                        settings.unsafe_fixes,
```
Comment (charliermarsh) [inline crates/ruff_cli/src/commands/check.rs:130]: I think accessing from `settings` here could lead to inconsistencies, since the Printer and Emitter and such will use the "global" value, but accessing from `settings` here will read the `unsafe-fixes` from a nested configuration file -- if I understand correctly.

For example... if you had `unsafe-fixes = true` in a `pyproject.toml` at the top of the project, and then `unsafe-fixes = false` in a `pyproject.toml` in a subdirectory, wouldn't the CLI _think_ you ran with `unsafe-fixes`, where internally you would _not_ run with `unsafe-fixes` when analyzing files in the subdirectory?

I suspect this should be passed down from the top-level like `fix_mode` for consistency throughout the execution.

**Review by charliermarsh — APPROVED** [review]
Nice! I have one question (or concern, depending on the response) but overall looks great.

File: crates/ruff_cli/src/commands/check.rs (hunk 25 above)
Comment (zanieb) [reply crates/ruff_cli/src/commands/check.rs:130]: Hm okay let me investigate that

File: crates/ruff_cli/src/diagnostics.rs (hunk 24 above)
Comment (zanieb) [reply crates/ruff_cli/src/diagnostics.rs:257]: 👍 I'll do that separately. I want to use `match` instead of `if matches! / else` just outside this too.

File: crates/ruff_cli/src/args.rs (hunk 23 above)
Comment (zanieb) [reply crates/ruff_cli/src/args.rs:81]: Hm.. you were the one that told me unsafe is an all or nothing concept haha I'll ponder this as well.

Should it be "potentially incorrect fixes"? Fwiw I feel like `--unsafe-fixes` is the place for a more detailed description.

Comment (charliermarsh) [reply crates/ruff_cli/src/args.rs:81]: Haha. I guess the way I thought about it was that a given fix is either safe or not. The collection of fixes may thereby contain unsafe changes.

File: crates/ruff_cli/src/commands/check.rs (hunk 25 above)
Comment (zanieb) [reply crates/ruff_cli/src/commands/check.rs]: Yeah you're right! Good catch — I misunderstood that call hierarchy.
